yt/__init__.py

```diff
@@ -9,10 +9,9 @@
 """
 import sys
 
-if sys.version_info[0] < 3:
-    raise Exception(
-        "Python 2 no longer supported.  Please install Python 3 for use with yt."
-    )
+py_version = sys.version_info
+if not (py_version.major == 3 and py_version.minor >= 7):
+    raise sys.exit("yt requires Python 3.7 or later.")
```

Comment:
I support switching our testing infra to not explicitly test against python 3.6, but I am not sure we should have a hard enforcement here. What about issuing a warning for python < 3.7 that we don't explicitly support those versions and users should expect odd behavior or something?

Reply:
Ok so I should have stated it earlier: the point is to make our lives easier when we start adding type annotations since there level of support is a bit limited in 3.6
I think it's preferable to have a clean PR that breaks compatibility on purpose, as part of a cycle, so it's clear to everyone we can start supporting 3.9 AND don't have to support 3.6 any longer.

Reply:
Right, I understand that. I just don't know why we can't choose to say "use at your own risk. This Python version is not explicitly supported in this version of yt" and then users can try to install and use yt with 3.6 if they want to, rather than issuing this more aggressive enforcement.

Reply:
I don't think it can be done in a way that the message would actually show up before a syntaxError is raised. Otherwise I would support this.

Reply:
Huh, I didn't realize that we used py 3.7 syntax.  Where do we use that?

My preference would be that until we actually move forward with the type annotation syntax that *isn't* supported, we not *hard* cancel for py3.6.  I say this because we have an unknown set of folks using py36, which was the install default for a long time, and we don't want to have a huge barrier to entry, since upgrading the python version can be confusing.  Plus, we might have folks using it with supercomputer python installs, and I think we need to err on the side of forgiveness there.

One thing we *should* do -- which we are several years behind on doing! -- is take a survey of the community to find out, among other things, what the python version distribution is.

Reply:
> Huh, I didn't realize that we used py 3.7 syntax. Where do we use that?

To be clear: we don't, atm.
I should add that I didn't write this PR with the intention of getting it merged early, it could very well sit here unmerged right until the moment we decide to use Python 3.7+ stuff, such as type annotations with decent support, but I wanted this to be ready when that day comes.
Personally I believe yt 4.0 should still support Python 3.6 but we should be allowed to break it for the following release.

Such a survey would be incredibly useful Matt !

Reply:
I agree that 4.0 should still support 3.6 -- once we up the release cadence I won't be as worried about dropping python versions but I do worry that dropping 3.6 for 4.0 could cause problems for our users. I also agree with matt that erring on the side of forgiveness for users and not hard cancelling a specific version. I think we need to find a solution that will be forgiving to users.

Reply:
Agreed. To sum up: this PR will be necessary at some point and contains (or should contain) every change needed _when_ we're start annotating code (or using other py3.7+ stuff).
At _that_ point, python 3.6's interpreter won't be able to run `import yt`, so I don't think it will be possible to transition with a "use at your own risks" warning.
In the mean time, no rush, we can keep python 3.6 as our oldest supported version.